op_join.go

```diff
@@ -23,9 +24,66 @@ func (JoinOperator) Phase() OperatorPhase {
 	return EvalPhase
 }
 
-// Dependencies ...
-func (JoinOperator) Dependencies(_ *Evaluator, _ []*Expr, _ []*tree.Cursor) []*tree.Cursor {
-	return []*tree.Cursor{}
+// Dependencies returns the nodes that (( join ... )) requires to be resolved
+// before its evaluation. Returns no dependencies on error, because who cares
+// about eval order if Run is going to bomb out anyway.
+func (JoinOperator) Dependencies(ev *Evaluator, args []*Expr, _ []*tree.Cursor) []*tree.Cursor {
+	DEBUG("Calculating dependencies for (( join ... ))")
+	deps := []*tree.Cursor{}
+	if len(args) < 2 {
+		DEBUG("Not enough arguments to (( join ... ))")
+		return []*tree.Cursor{}
+	}
+
+	//skip the separator arg
+	for _, arg := range args[1:] {
+		if arg.Type == Literal {
+			continue
+		}
+		if arg.Type != Reference {
+			DEBUG("(( join ... )) argument not Literal or Reference type")
+			return []*tree.Cursor{}
```

Comment:
It is early. I did not read well.

> On Sep 12, 2016, at 9:24 AM, Thomas Mitchell <notifications@github.com> wrote:
> 
> In op_join.go <https://github.com/geofffranks/spruce/pull/165#discussion_r78370885>:
> 
> > +func (JoinOperator) Dependencies(ev *Evaluator, args []*Expr, _ []*tree.Cursor) []*tree.Cursor {
> > +	DEBUG("Calculating dependencies for (( join ... ))")
> > +	deps := []*tree.Cursor{}
> > +	if len(args) < 2 {
> > +		DEBUG("Not enough arguments to (( join ... ))")
> > +		return []*tree.Cursor{}
> > +	}
> > +
> > +	//skip the separator arg
> > +	for _, arg := range args[1:] {
> > +		if arg.Type == Literal {
> > +			continue
> > +		}
> > +		if arg.Type != Reference {
> > +			DEBUG("(( join ... )) argument not Literal or Reference type")
> > +			return []*tree.Cursor{}
> It stops only if the arg is not a Literal OR a reference, which is rightfully an error. That said, I can add a test case for this
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/geofffranks/spruce/pull/165/files/4c4e7ab9d901e79fb5ed98480c5798e841f0dfa3#r78370885>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AG2SUlUd9pRKzfWUam81t7g2pPPUg-Miks5qpVJ8gaJpZM4J3XjJ>.
>